WIP: Audio recording and playback

dpgeorge opened the pull request:
This PR adds audio/microphone recording and playback capabilities, as per the docs proposal: https://github.com/bbcmicrobit/micropython/pull/791

It currently supports:
- `audio.AudioFrame(size)` to create large audio frames
- `audio.play(audio_frame)` can now take a single `AudioFrame` to play
- `audio.set_rate(rate)` can set the playback rate on the fly
- `audio.sound_level()` returns the current sound level being played
- `microphone.record_into(buffer, rate, wait)` to record audio
- `microphone.is_recording()` to check if recording is ongoing
- `microphone.stop_recording()` to stop recording

There is a test program called `src/test_record.py` which uses all the above to show how it works.

dpgeorge committed:
codal_port/modaudio: Add audio.sound_level() method.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Extend AudioFrame constructor to take opt size.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Rework audio playing to play a long AudioFrame.

- Audio playback now supports AudioFrames with arbitrary size.
- Allows passing a single AudioFrame to audio.play.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Remove buffer expansion.

Playing generated samples (eg from v1 code) now sounds better without
buffer expansion.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_app: Add microphone recording interface.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/microbit_microphone: Add methods to record.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
src: Add test recording program.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
@microbit-carlos Please try out the test program and see how it goes.

dpgeorge committed:
codal_port/modaudio: Make AudioFrame constructor public.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Rename "size" field to "alloc_size".

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_app/microbithal_microphone: Store the current recording length.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Add a "used_size" field to AudioFrame.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Add rate to AudioFrame.

Make the constructor take duration and rate arguments, and add get_rate() /
set_rate() methods.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/microbit_microphone: Use rate from AudioFrame, or update it.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Use AudioFrame rate as rate when playing it.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
src: Update test_record.py.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I have updated this PR with the following additions/changes:
- added `microphone.record(duration, rate=7812)`
- added a "used length" member to `AudioFrame`, so that it remembers how much valid recording there is; this value is set when recording, and when playing back only that much data is played
- added a "rate" member to `AudioFrame`, and `set_rate()` / `get_rate()` members
- calling `microphone.record_into()` will either use the rate of the given `AudioFrame`, or set the rate on that `AudioFrame`, if the rate is passed into this record method
- playing an `AudioFrame` will play at its rate, and changing the rate (via `audio_frame.set_rate()`) while it's playing will dynamically adjust the playback rate
- the `AudioFrame` constructor is changed to `AudioFrame(duration=-1, rate=7812)`, so you can just pass in a duration in the simple case eg `AudioFrame(2000)`

dpgeorge committed:
codal_app/main: Set microphone gain to 0.2.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Take the sqrt of sound_level.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Mostly use "alloc_size" instead of "used_size".

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Round up AudioFrame size to nearest 32.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Increas "used_size" when data is written.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
src: Update test_record.py.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Update to build with latest micropython.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Some tweaks have been made to how the "used_size" entry in `AudioFrame` is used/set, the default microphone gain set to 0.2, and this PR has been rebased on the latest master which includes an update to the latest MicroPython version.

@microbit-carlos this is now ready for wider testing.

microbit-carlos commented:
Thanks Damien!
Btw,  the next CODAL tag has been released yesterday, which should resolve some of the issues with changing the sampling rate to non-proportional values:
https://github.com/lancaster-university/codal-microbit-v2/releases/tag/v0.2.65
Diff: https://github.com/lancaster-university/codal-microbit-v2/compare/v0.2.63...v0.2.65

dpgeorge committed:
codal_app/microbithal_microphone: Make input streaming use pullInto.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
@microbit-carlos I have now updated this PR to use CODAL v0.2.66.  It's working well.

microbit-carlos commented on src/codal_port/microbit_microphone.c:
```diff
@@ -147,13 +148,89 @@ STATIC mp_obj_t microbit_microphone_get_events(mp_obj_t self_in) {
+    }
+
+    // Start the recording.
+    microbit_hal_microphone_start_recording(audio_frame->data, audio_frame->alloc_size, &audio_frame->used_size, audio_frame->rate);
+
+    return mp_const_none;
```
I think this is missing the wait flag implementation.

For example, with this bit of code, the display shows "R" inmediately:
```python
my_recording = audio.AudioFrame(duration=3000)
microphone.record_into(my_recording, wait=True)
display.show("R")
```

dpgeorge replied:
`wait` is now implemented.

microbit-carlos commented:
With this example the playback rate doesn't seem to change for playback.

```python
from microbit import *

RECORDING_SAMPLING_RATE = 7812

while True:
    if pin_logo.is_touched():
        # Record and play back at the same rate
        my_recording = microphone.record(duration=3000, rate=RECORDING_SAMPLING_RATE)
        audio.play(my_recording)
        del my_recording

    if button_a.is_pressed():
        # Play back at half the sampling rate
        my_recording = microphone.record(duration=3000, rate=RECORDING_SAMPLING_RATE)
        audio.set_rate(RECORDING_SAMPLING_RATE // 2)
        audio.play(my_recording)
        del my_recording

    if button_b.is_pressed():
        # Play back at twice the sampling rate
        my_recording = microphone.record(duration=3000, rate=RECORDING_SAMPLING_RATE)
        audio.set_rate(RECORDING_SAMPLING_RATE * 2)
        audio.play(my_recording)
        del my_recording

    sleep(200)
```

microbit-carlos commented:
Also, I think the `microphone.set_sensitivity(gain)` still needs to be implemented.

microbit-carlos commented:
Another small one, if the AudioFrame rate is set to a negative number it overflows the calculation (well, maybe it's converting it to an unsigned int at some point):

```
>>> AudioFrame(1000, -1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError: memory allocation failed, allocating 4294990 bytes
```

microbit-carlos commented:
Looks like `set_rate()` overflows:

```
>>> af = AudioFrame()
>>> af.set_rate(2**16 - 1)
>>> af.get_rate()
65535
>>> af.set_rate(2**16)
>>> af.get_rate()
0
>>> af.set_rate(2**16 + 1)
>>> af.get_rate()
1
```

And on that note, should the rate be limited to 16 bits? Should it be increased to 32?

microbit-carlos commented:
Should `copyfrom()` also duplicate the length from the source?
Before these changes AudioFrame was always 32 bytes, but now we can have different size AudioFrames.

I was also wondering if we should also make it a static function to be able to do `new = AudioFrame.copyfrom(old)`.
Otherwise we now we have to figure out a way to match the length of the `old` instance somehow when creating the `new`, as we cannot even do something like `new = AudioFrame(size=len(old)); new.copyfrom(old)`.

microbit-carlos commented:
Does the AudioFrame returned by `microphone.record()` purposely does not have a size that is a multiple of 32 bytes, or is `len(AudioFrame())` returning the "used size" instead of the "total size"?

```
>>> len(microphone.record(duration=1000))
7812
>>> 7812 / 32
244.125
```

microbit-carlos commented:
Passing a negative value to the `microphone.record(-1)` seems to lock up and never return.

dpgeorge committed:
codal_port/microbit_microphone: Implement wait argument to record_into.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_app/microbithal_microphone: Remove CODAL workaround.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> Ah! the type was that this example was using `audio.set_rate()`, which I guess we need to remove from this branch?

I have now removed `audio.set_rate()`.

dpgeorge commented:
> I think the `microphone.set_sensitivity(gain)` still needs to be implemented.

Yes, correct, it still needs to be implemented.

Would this just call CODAL's `uBit.audio.processor->setGain()`?  If so, what is the argument to `set_sensitivity()`, is it a float or integer, and is the value inverted before passing through to `setGain()`?  And what values should we use for the `SENSITIVITY_xxx` constants?

dpgeorge commented:
> Another small one, if the AudioFrame rate is set to a negative number it overflows the calculation

This was a mistake, it was supposed to be checking for negative rates.  Now fixed.

> Looks like `set_rate()` overflows/converts as well:

I also fixed this, and increased the internal rate variable to 32-bits (I thought 16-bit would be enough for the rate, but it doesn't cost anything to make it 32-bits).

microbit-carlos commented:
> Would this just call CODAL's `uBit.audio.processor->setGain()`? If so, what is the argument to `set_sensitivity()`, is it a float or integer, and is the value inverted before passing through to `setGain()`? And what values should we use for the `SENSITIVITY_xxx` constants?

Yes, only need to call the `uBit.audio.processor->setGain()` and the values are floats and we can use the same three levels as [MakeCode](https://github.com/microsoft/pxt-microbit/blob/2dfa9fac50f4f677348d1fb16b2872b4045beb97/libs/audio-recording/recording.ts#L171-L179):
- `SENSITIVITY_LOW` = 0.079
- `SENSITIVITY_MEDIUM` = 0.2 (the default)
- `SENSITIVITY_HIGH` = 1.0

microbit-matt-hillsdon commented:
The existing sim implementation of playing AudioBuffers has always been problematic because of the 4ms chunk size. For audio frames of longer duration I had more hope, but at the moment the chunk size used is the same even though we likely have a much larger buffer in the frame itself.

Our work-in-progress record/playback sim implementation works OK if you increase LOG_AUDIO_CHUNK_SIZE from 5 to 6. (it's possible that on slower computers a bigger buffer still might help, I've not had a chance to test yet). We can't just do that as I think it means regular audio frames are pitch/rate shifted but it might show a way forward for the sim.

The sim changes without the LOG_AUDIO_CHUNK_SIZE change can be seen here:
https://review-python-simulator.usermbit.org/beta-updates/demo.html (https://github.com/microbit-foundation/micropython-microbit-v2-simulator/pull/113) - see sample "Record".  For now you'll have to rebuild locally to see the benefit of the buffer size change. We've not yet looked at edge cases etc. but I think this problem will remain.

dpgeorge committed:
codal_app/microbithal_microphone: Add func to set the sensitivity.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/microbit_microphone: Add set_sensitivity method and consts.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Ensure AudioFrame size is non-zero.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/microbit_microphone: Validate duration and rate args.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
The following changes have been made to this branch/PR:
- Added `microphone.set_sensitivity(gain)` along with the constants `microphone.SENSITIVITY_LOW`, `SENSITIVITY_MEDIUM`, `SENSITIVITY_HIGH`.
- Round up `AudioFrame` size to 32 bytes when created by `microphone.record()`.
- Raise exceptions if `duration` or `rate` arguments to `microphone.record()` are not positive.

dpgeorge commented:
> Should `copyfrom()` also duplicate the length from the source? 

This needs discussion.  `copyfrom()` actually allows copying from any object with the buffer protocol (eg bytes, memoryview) and they won't necessarily have the `used_size` attribute.

This method should probably just set `used_size` to the total bytes copied fro the buffer-like object.  The only issue there is copying from another `AudioFrame` it will copy the total allocated size, not the `used_size`, of the other `AudioFrame`.

> I was also wondering if we should also make it a static function to be able to do `new = AudioFrame.copyfrom(old)`. Otherwise we now we have to figure out a way to match the length of the `old` instance somehow when creating the `new`, as we cannot even do something like `new = AudioFrame(size=len(old)); new.copyfrom(old)`.

How about allowing passing an `AudioFrame` to the `AudioFrame` constructor?  This is how `list`, `dict` etc work in Python.  (Did we already discuss point?)

microbit-carlos commented on src/codal_port/microbit_microphone.c:
```diff
@@ -147,13 +155,111 @@ STATIC mp_obj_t microbit_microphone_get_events(mp_obj_t self_in) {
+    { MP_ROM_QSTR(MP_QSTR_stop_recording), MP_ROM_PTR(&microbit_microphone_stop_recording_obj) },
+
+    // Constants.
+    { MP_ROM_QSTR(MP_QSTR_SENSITIVITY_LOW), MP_ROM_INT(0.079) },
+    { MP_ROM_QSTR(MP_QSTR_SENSITIVITY_MEDIUM), MP_ROM_INT(0.2) },
+    { MP_ROM_QSTR(MP_QSTR_SENSITIVITY_HIGH), MP_ROM_INT(1.0) },
```
These should be floats:
- https://github.com/microbit-foundation/micropython-microbit-v2/issues/204

dpgeorge replied:
Now fixed.

microbit-carlos commented:
> > Should `copyfrom()` also duplicate the length from the source?
> 
> This needs discussion. `copyfrom()` actually allows copying from any object with the buffer protocol (eg bytes, memoryview) and they won't necessarily have the `used_size` attribute.
> 
> This method should probably just set `used_size` to the total bytes copied fro the buffer-like object. The only issue there is copying from another `AudioFrame` it will copy the total allocated size, not the `used_size`, of the other `AudioFrame`.

This conversation has already been moved to:
- https://github.com/microbit-foundation/micropython-microbit-v2/issues/190



> > I was also wondering if we should also make it a static function to be able to do `new = AudioFrame.copyfrom(old)`. Otherwise we now we have to figure out a way to match the length of the `old` instance somehow when creating the `new`, as we cannot even do something like `new = AudioFrame(size=len(old)); new.copyfrom(old)`.
> 
> How about allowing passing an `AudioFrame` to the `AudioFrame` constructor? This is how `list`, `dict` etc work in Python. (Did we already discuss point?)

We can move this one to:
- https://github.com/microbit-foundation/micropython-microbit-v2/issues/189

dpgeorge committed:
codal_port/modaudio: Make None the default to AudioFrame().

And allow duration to be a float, for more accurate sizing.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Allow AudioFrame to be arbitrary length.

No longer needs to be a multiple of 32.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/microbit_microphone: Require rate to be positive.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Fix typecode of AudioFrame buffer to be unsigned.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Use mp_sched_schedule_node for audio fetcher.

This is more efficient than mp_sched_schedule(), and can never fail.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Separate default AudioFrame and output buffer size.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Allow output buffer to be larger than 32 bytes.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Stop streaming audio when data is exhausted.

Signed-off-by: Damien George <damien@micropython.org>

zazer0 commented:
Hi, is this currently usable? Keen to use my microbit v2 as microphone in + audio out for an art project if possible :)

dpgeorge commented:
> Hi, is this currently usable?

Yes this branch is usable.  We are currently fine-tuning the Python-level API (function names and behaviour) so things will change, but the overall functionality will remain.

dpgeorge committed:
codal_port: Implement AudioTrack and AudioRecording.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
src: Update test_record.py.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Remove used_size entry from AudioFrame type.

It's no longer needed now that there is AudioTrack.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Make AudioFrame add and mult helpers public.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/microbit_audiotrack: Implement +,+=,-,-=,*,*= on AudioTrack.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
src: Update test_record.py to use mult for volume.

Signed-off-by: Damien George <damien@micropython.org>

microbit-matt-hillsdon commented on src/test_record.py:
```diff
@@ -0,0 +1,53 @@
+        display.clear()
+        my_track *= 2  # amplify volume
+    if button_b.is_pressed():
+        audio.play(my_track, wait=False)
```
This line fails with 

```
Traceback (most recent call last):
  File "main.py", line 40, in <module>
NameError: name 'my_track' isn't defined
```

if you push B before A. This flagged by the basic type checking that we have enabled in the Python Editor so would be good to avoid if we can.

dpgeorge replied:
Good point.  Now fixed.

dpgeorge committed:
src: In test_record.py, make sure my_track is defined.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/microbit_audiorecording: Make track() args positional.

Signed-off-by: Damien George <damien@micropython.org>

microbit-matt-hillsdon commented on src/test_record.py:
```diff
@@ -0,0 +1,54 @@
+            sleep(50)
+        display.clear()
+        my_track *= 2  # amplify volume
+    if button_b.is_pressed() and my_track:
```
I think playing is triggered more reliably with was_pressed here, at least in the sim.

dpgeorge replied:
Yes, good idea, now changed.

dpgeorge committed:
src: Use was_pressed in test_record.py.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Improve feeding of audio pipeline.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_app/microbithal_audio: Add microbit_hal_audio_is_playing() func.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Make audio waiting wait for audio to be silent.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
codal_port/modaudio: Make sure final partial audio data is sent out.

Signed-off-by: Damien George <damien@micropython.org>
